pkg/activator/throttler.go

```diff
@@ -0,0 +1,151 @@
+package activator
+
+import (
+	"github.com/knative/serving/pkg/queue"
+	"go.uber.org/zap"
+	"sync"
+	"github.com/knative/serving/pkg/client/informers/externalversions/serving/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	v1alpha12 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"errors"
+	"strings"
+	corev1 "k8s.io/api/core/v1"
+	"fmt"
+)
+
+// NewThrottler creates a new Throttler
+func NewThrottler(queueDepth, maxConcurrency int32, revisionInformer v1alpha1.RevisionInformer, logger *zap.SugaredLogger) *Throttler {
+	breakers := make(map[RevisionID]queue.Breaker)
+	return &Throttler{breakers: breakers, queueDepth: queueDepth, maxConcurrency: maxConcurrency, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler allows to update the max concurrency of registered Breakers
+type Throttler struct {
+	breakers         map[RevisionID]queue.Breaker
+	revisionInformer v1alpha1.RevisionInformer
+	queueDepth       int32
+	maxConcurrency   int32
+	logger           *zap.SugaredLogger
+	mux              sync.Mutex
+}
+
+// Add potentially creates a new breaker for the revision
+// it returns true if the breaker already exists, false otherwise
+func (t *Throttler) Add(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	if _, ok := t.breakers[rev]; !ok {
+		breaker := queue.NewBreaker(t.queueDepth, t.maxConcurrency, 0)
+		t.breakers[rev] = *breaker
+		return ok
+	}
+	return true
+}
+
+// Exists returns true if a Breaker already exists for the given revision
+func (t *Throttler) Exists(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, ok := t.breakers[rev]
+	return ok
+}
+
+// Removes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	var containerConcurrency int32
+	// Knative Revision and k8s service have different names
+	baseName := strings.Split(rev.Name, "-service")[0]
```

Comment:
This seems dangerous, as the naming convention can change any time. If we need to do string mangling here, we should put a method in `revison/resources/names.go` that transforms a service name to a revision name for us. Or maybe simpler: We can generate a map from service -> revision name once a revision is added and use that to look the revision up?

Reply:
I think they could be mutually inclusive. Both less string concatenation and putting all renaming in one place make sense